libs/langchain/langchain/schema/runnable/locals.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+from typing import Any, AsyncIterator, Dict, Iterator, Mapping, Optional, Union
+
+from langchain.callbacks.manager import (
+    AsyncCallbackManagerForChainRun,
+    CallbackManagerForChainRun,
+)
+from langchain.load.serializable import Serializable
+from langchain.schema.runnable.base import Input, Output, Runnable
+from langchain.schema.runnable.config import RunnableConfig
+from langchain.schema.runnable.passthrough import RunnablePassthrough
+
+
+class PutLocalVar(RunnablePassthrough):
+    key: Union[str, Mapping[str, str]]
+    """The key(s) to use for storing the input variable(s) in local state.
+    
+    If a string is provided then the entire input is stored under that key. If a 
+        Mapping is provided, then the map values are gotten from the input and 
+        stored in local state under the map keys.
+    """
+
+    def __init__(self, key: Union[str, Mapping[str, str]], **kwargs: Any) -> None:
+        super().__init__(key=key, **kwargs)
+
+    def _concat_put(
+        self,
+        input: Input,
+        *,
+        config: Optional[RunnableConfig] = None,
+        replace: bool = False,
+    ) -> None:
+        if config is None:
+            raise ValueError(
+                "PutLocalVar should only be used in a RunnableSequence, and should "
+                "therefore always receive a non-null config."
+            )
+        if isinstance(self.key, str):
+            if self.key not in config["_locals"] or replace:
+                config["_locals"][self.key] = input
+            else:
+                config["_locals"][self.key] += input
+        elif isinstance(self.key, Mapping):
+            if not isinstance(input, Mapping):
+                raise TypeError(
+                    f"Received key of type Mapping but input of type {type(input)}. "
+                    f"input is expected to be of type Mapping when key is Mapping."
+                )
+            for input_key, put_key in self.key.items():
+                if put_key not in config["_locals"] or replace:
+                    config["_locals"][put_key] = input[input_key]
+                else:
+                    config["_locals"][put_key] += input[input_key]
+        else:
+            raise TypeError(
+                f"`key` should be a string or Mapping[str, str], received type "
+                f"{(type(self.key))}."
+            )
+
+    def invoke(self, input: Input, config: Optional[RunnableConfig] = None) -> Input:
+        self._concat_put(input, config=config, replace=True)
+        return super().invoke(input, config=config)
+
+    async def ainvoke(
+        self,
+        input: Input,
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Optional[Any],
+    ) -> Input:
+        self._concat_put(input, config=config, replace=True)
```

Comment:
should this be using `_concat_put` implementation? it's not a streaming variant

Reply:
see replace=True. This was to avoid duplicating the implementation, and thus potentially introducing diffs in behaviour between stream and non stream

Reply:
maybe just _put?